text/0000-storage-proof-size-hostfunction.md

```diff
@@ -0,0 +1,48 @@
+| **Start Date**  | 30 October 2023                                                                    |
+| **Description** | Host function to provide the storage proof size to runtimes.                                                                    |
+| **Authors**     | Sebastian Kunert                                                                                            |
+
+## Summary
+
+This RFC proposes a new host function for parachains, `storage_proof_size`.  It shall provide the size of the currently recorded storage proof to the runtime. Runtime authors can use the proof size to improve block utilization by retroactively reclaiming unused storage weight.
+
+## Motivation
+The number of extrinsics that are included in a parachain block is limited by two constraints: execution time and proof size. FRAME weights cover both concepts, and block-builders use them to decide how many extrinsics to include in a block. However, these weights are calculated ahead of time by benchmarking on a machine with reference hardware. The execution-time properties of the state-trie and its storage items are unknown at benchmarking time. Therefore, we make some assumptions about the state-trie:
+- **Trie Depth:** We assume a trie depth to account for intermediary nodes.
+- **Storage Item Size:** We make a pessimistic assumption based on the `MaxEncodedLen` trait.
+
+These pessimistic assumptions lead to an overestimation of storage weight, negatively impacting block utilization on parachains.
+
+In addition, the current model does not account for multiple accesses to the same storage items. While these repetitive accesses will not increase storage-proof size, the runtime-side weight monitoring will account for them multiple times. Since the proof size is completely opaque to the runtime, we can not implement retroactive storage weight correction.
+
+A solution must provide a way for the runtime to track the exact storage-proof size consumed on a per-extrinsic basis.
+
+## Stakeholders
+- **Parachain Teams:** They should include this host function in their runtime and node.
+- **Light-client Implementors:** They should include this host function in their runtime and node.
+
+## Explanation
+This RFC proposes a new host function that exposes the storage-proof size to the runtime. As a result, runtimes can implement storage weight reclaiming mechanisms that improve block utilization.
+
+This RFC proposes the following host function signature:
+```rust
+fn storage_proof_size() -> u64;
+```
+The host function MUST return an unsigned 64-bit integer value representing the current proof size. In block-execution and block-import contexts, this function MUST return the current size of the proof. To achieve this, parachain node implementors need to enable proof recording for block imports. In contexts without proof recording, this function MUST return 0. 
```

Comment:
> In contexts without proof recording, this function MUST return 0. 

This would introduce non-determinism, unless you specify in which situations proof recording is enabled and which situations it is disabled.

Reply:
It's deterministic for polkadot verification though.

Yeah, parachains could screw up light client protocols, or even their own block building, by doing this, but these cases seem like "parachain authors fucked up".  Anyways, they could already do this by using `static mut`s, which they must do sometimes given some design choices in Frame.

Reply:
> IMO the only way that makes sense to me would be to always unconditionally enable proof recording

Hmm yes I see, the downside of this is that it comes with a performance impact.

Wdyt about adjusting the wording to the opposite, i.e. the host function must always return 0 except for block building and import, regardless of proof recording setting?

Reply:
> Wdyt about adjusting the wording to the opposite, i.e. the host function must always return 0 except for block building and import, regardless of proof recording setting?

I guess that works.

However if I remember correctly `BlockBuilder_apply_extrinsic` is used when generating a block but also outside of block building to know whether a transaction will succeed (which is an information that validating a transaction alone doesn't provide), so the implementation of this function would need to make sure to not accidentally say that transactions are invalid if that function returns 0.

It's not a blocker, but it's an extra footgun that wouldn't be there if proof record was always enabled.